src/System.IO.Pipelines.Networking.Tls/SecurePipeline.cs

```diff
@@ -0,0 +1,255 @@
+﻿using System.Threading.Tasks;
+
+namespace System.IO.Pipelines.Networking.Tls
+{
+    public class SecurePipeline<T> : ISecurePipeline where T : ISecureContext
+    {
+        private readonly IPipelineConnection _lowerConnection;
+        private readonly PipelineReaderWriter _outputPipeline;
+        private readonly PipelineReaderWriter _inputPipeline;
+        private readonly T _contextToDispose;
+        private TaskCompletionSource<ApplicationProtocols.ProtocolIds> _handShakeCompleted;
+
+        internal SecurePipeline(IPipelineConnection inConnection, PipelineFactory pipelineFactory, T secureContext)
+        {
+            _contextToDispose = secureContext;
+            _lowerConnection = inConnection;
+            _inputPipeline = pipelineFactory.Create();
+            _outputPipeline = pipelineFactory.Create();
+            StartWriting();
+        }
+
+        public IPipelineReader Input => _outputPipeline;
+        public IPipelineWriter Output => _inputPipeline;
+        public CipherInfo CipherInfo => _contextToDispose.CipherInfo;
+
+        public Task<ApplicationProtocols.ProtocolIds> PerformHandshakeAsync()
+        {
+            return _handShakeCompleted?.Task ?? DoHandShake();
+        }
+
+        private async Task<ApplicationProtocols.ProtocolIds> DoHandShake()
```

Comment:
The public method was "Handshake" this one is "HandShake".  The former is probably preferred.